### ISSUE TITLE
Problems with TC-Powercore

### ISSUE THREAD

**Issue by Joerg69** [teragonaudio/MrsWatson#201]
Hi there!

I've tried to run MrsWatson with my TC-powercore - unfortunaltly it's not working. MrsWatson works fine with local DLLs bur everytime I want to use a DLL that requires the support from the powercore MrsWatson crashes!

Here are her "last words":

C:\Users\Joerg\Desktop>mrswatson -p C:\Users\Joerg\Desktop\tubifex.dll -i C:\use
rs\joerg\desktop\musik.wav -o c:\users\joerg\desktop\musik-out.wav --realtime
- 00000000 000000 MrsWatson version 0.9.8 initialized, build 20150122
- 00000000 000000 Setting 2 channels
- 00000000 000000 Setting sample rate to 44100Hz
- 00000000 000000 Plugin 'C:\Users\Joerg\Desktop\tubifex.dll' is of type VST2.x
- 00000000 000000 Opening VST2.x plugin 'C:\Users\Joerg\Desktop\tubifex.dll'
W 00000000 001107 Call to deprecated function 'audioMasterNeedIdle' made by plug
in 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterNeedIdle' made by plug
in 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001107 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001123 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001123 Call to deprecated function 'audioMasterNeedIdle' made by plug
in 'Vnix'
W 00000000 001123 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001123 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001123 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001123 Call to deprecated function 'audioMasterPinConnected' made by
plugin 'Vnix'
W 00000000 001139 Call to deprecated function 'audioMasterNeedIdle' made by plug
in 'Vnix'
W 00000000 001139 Call to deprecated function 'audioMasterNeedIdle' made by plug
in 'Vnix'
W 00000000 001139 Call to deprecated function 'audioMasterNeedIdle' made by plug
in 'Vnix'
- 00000000 001139 Starting processing input source
W 00001024 001201 Short write occurred while writing samples
INTERNAL ERROR: framesProcessed (1539) != getAudioClock()->currentFrame (1536)
  This should not have happened. Please take a minute to report a bug.
  Support website: https://github.com/teragonaudio/mrswatson/issues
  Support email: support@teragonaudio.com
INTERNAL ERROR: framesProcessed (2051) != getAudioClock()->currentFrame (2048)
  This should not have happened. Please take a minute to report a bug.
  Support website: https://github.com/teragonaudio/mrswatson/issues
  Support email: support@teragonaudio.com
INTERNAL ERROR: framesProcessed (2563) != getAudioClock()->currentFrame (2560)
  This should not have happened. Please take a minute to report a bug.
  Support website: https://github.com/teragonaudio/mrswatson/issues
  Support email: support@teragonaudio.com
INTERNAL ERROR: framesProcessed (3075) != getAudioClock()->currentFrame (3072)
  This should not have happened. Please take a minute to report a bug.
  Support website: https://github.com/teragonaudio/mrswatson/issues
  Support email: support@teragonaudio.com

Any Idea???

**Comment by nikreiman** [comment]
The internal error shown is a known issue, but shouldn't effect the program or cause it to crash. However, I'm curious to know if the short write still occurs when using an output source like PCM (just use `-o out.pcm` instead). This could be related to the new audiofile support.

It seems that the crash occurs sometime after the first few blocks have been processed. Unfortunately, without a stacktrace it is almost impossible to know what happened. I did a bit of searching around, and it seems that if you install the windows debugging tools and run the program through windbg or procdump (see http://stackoverflow.com/a/24209279/14302 ), then you can get a stacktrace.

This is probably a bit more work than should be needed though, I will try to integrate crashrpt (https://code.google.com/p/crashrpt/) in MrsWatson instead. I have created a separate issue for this, see #204.

**Comment by JoergKn** [comment]
Dear Nik!

 

Thanks for the feedback and spending your time on improving this great program!! 

 

First of all I have to confess that I was wrong with my statement that MrsWatson is crashing – in fact it “only” produces a very large amount of errors messages so I’ve canceled it after a while without let it run through. When I let it run through it produces a file with the VST effect included but unfortunately the amount of samples in the output file is different!

 



 

When looking into the output file there is a time shift of about 950 samples (see picture enclosed).

 

I already had a look at your sources and I found the way you handle the Plug-In latency is very complicated by moving the pointer by the amount of samples from the latency. I’m pretty sure that this is the point where something will be messed up after a few cycles. Wouldn’t it be easier to process the complete file first and in a second step cut of the latency at the beginning from the finished file!?? 

 

All the best to you and happy bug hunting!

 

Joerg

 

 

Von: Nik Reiman [mailto:notifications@github.com] 
Gesendet: Freitag, 3. April 2015 16:32
An: teragonaudio/MrsWatson
Cc: JoergKn
Betreff: Re: [MrsWatson] Problems with TC-Powercore (#201)

 

The internal error shown is a known issue, but shouldn't effect the program or cause it to crash. However, I'm curious to know if the short write still occurs when using an output source like PCM (just use -o out.pcm instead). This could be related to the new audiofile support.

It seems that the crash occurs sometime after the first few blocks have been processed. Unfortunately, without a stacktrace it is almost impossible to know what happened. I did a bit of searching around, and it seems that if you install the windows debugging tools and run the program through windbg or procdump (see http://stackoverflow.com/a/24209279/14302 ), then you can get a stacktrace.

This is probably a bit more work than should be needed though, I will try to integrate crashrpt (https://code.google.com/p/crashrpt/) in MrsWatson instead. I have created a separate issue for this, see #204 <https://github.com/teragonaudio/MrsWatson/issues/204> .

—
Reply to this email directly or view it on GitHub <https://github.com/teragonaudio/MrsWatson/issues/201#issuecomment-89304464> .  <https://github.com/notifications/beacon/ALKEUws1mIcq8xveGOiCq8bduoxNjLVSks5n7pvPgaJpZM4D2dQd.gif> 

  _____  

E-Mail ist virenfrei.
Von AVG überprüft - www.avg.de
Version: 2015.0.5863 / Virendatenbank: 4321/9442 - Ausgabedatum: 03.04.2015

**Comment by JoergKn** [comment]
Dear Nik!

 

I’ve tried to change the output file to PCM and there are no more error messages coming up!!! Unfortunately I cannot check the PCM file as import to my program fails – I’ve tried to convert with SOX but it says: “Riff header not found” … ?

 

But I like to give you more precise information to help you finding the bug. As it is not easy to read latencies from a music file I’ve processed a sine burst and measured the difference between the first zero crossing (see picture enclosed). 

To explain the pictures: The input is on the top and below the output from MrsWatson. My Plug-In got a latency of 450 samples when read back the plug-in info.

 



 

You can see that the output is 961 samples shifted forward!

 

The file length of the processed file is 511 samples shorter than the original (see picture enclosed)!

 

Hope this help - have a nice Easter!

 

Joerg

 

 

 

Von: Joerg Knieschewski [mailto:mail@joerg-knieschewski.de] 
Gesendet: Freitag, 3. April 2015 23:14
An: 'teragonaudio/MrsWatson'
Betreff: AW: [MrsWatson] Problems with TC-Powercore (#201)

 

Dear Nik!

 

Thanks for the feedback and spending your time on improving this great program!! 

 

First of all I have to confess that I was wrong with my statement that MrsWatson is crashing – in fact it “only” produces a very large amount of errors messages so I’ve canceled it after a while without let it run through. When I let it run through it produces a file with the VST effect included but unfortunately the amount of samples in the output file is different!

 



 

When looking into the output file there is a time shift of about 950 samples (see picture enclosed).

 

I already had a look at your sources and I found the way you handle the Plug-In latency is very complicated by moving the pointer by the amount of samples from the latency. I’m pretty sure that this is the point where something will be messed up after a few cycles. Wouldn’t it be easier to process the complete file first and in a second step cut of the latency at the beginning from the finished file!?? 

 

All the best to you and happy bug hunting!

 

Joerg

 

 

Von: Nik Reiman [mailto:notifications@github.com] 
Gesendet: Freitag, 3. April 2015 16:32
An: teragonaudio/MrsWatson
Cc: JoergKn
Betreff: Re: [MrsWatson] Problems with TC-Powercore (#201)

 

The internal error shown is a known issue, but shouldn't effect the program or cause it to crash. However, I'm curious to know if the short write still occurs when using an output source like PCM (just use -o out.pcm instead). This could be related to the new audiofile support.

It seems that the crash occurs sometime after the first few blocks have been processed. Unfortunately, without a stacktrace it is almost impossible to know what happened. I did a bit of searching around, and it seems that if you install the windows debugging tools and run the program through windbg or procdump (see http://stackoverflow.com/a/24209279/14302 ), then you can get a stacktrace.

This is probably a bit more work than should be needed though, I will try to integrate crashrpt (https://code.google.com/p/crashrpt/) in MrsWatson instead. I have created a separate issue for this, see #204 <https://github.com/teragonaudio/MrsWatson/issues/204> .

—
Reply to this email directly or view it on GitHub <https://github.com/teragonaudio/MrsWatson/issues/201#issuecomment-89304464> .  <https://github.com/notifications/beacon/ALKEUws1mIcq8xveGOiCq8bduoxNjLVSks5n7pvPgaJpZM4D2dQd.gif> 

  _____  

E-Mail ist virenfrei.
Von AVG überprüft - www.avg.de
Version: 2015.0.5863 / Virendatenbank: 4321/9442 - Ausgabedatum: 03.04.2015

**Comment by nikreiman** [comment]
It looks like here the problem is with audiofile WAVE output (well, probably not the audiofile library itself, but rather how MW uses it). Otherwise, my only guess would be that plugin delay compensation is not working correctly in 0.9.8.

**Comment by jarl-dk** [comment]
I am no longer working with sound files, i.e. MW. However, @nikreiman , if you dig deeper and really suspect that the delay compensation is buggy let me know, I may learn something, I may help with something.

**Comment by JoergKn** [comment]
The new 32-bit windows build from today still got problems with the delay compensation!

I've tried the following: processed a wav-file with a sine burst, and one Plug-in with 450 frames internal delay and the same plug-in PLUS a plug-in with zero frames internal delay:

![mrswatson_shift_neu](https://cloud.githubusercontent.com/assets/11699283/7010278/25d7efdc-dca0-11e4-9b8a-fc5620988d05.PNG)

The upper is the original, the middle the one with one 450 frames and the lower one with zero frames Plug-in. With one Plug-in and 450 frames the signal is moved 960 frames backwards and with two Plug-ins another 539 frames - so in total 1499 frames!

![mrswatson_shift2_neu](https://cloud.githubusercontent.com/assets/11699283/7010398/252ef6e2-dca1-11e4-9232-7f683017ee14.PNG)

To my surprise the file length is ONLY changed when the 450 frames Plug-in will be processed:

![mrswatson_filelenght_neu](https://cloud.githubusercontent.com/assets/11699283/7010437/8594f1b2-dca1-11e4-80ab-a16017a4c9da.PNG)

With one 450 frames Plug-in the output file became 510 frames longer!

Hope you can fix this!

**Comment by jarl-dk** [comment]
@nikreiman : I believe the buffer delay compensation has been broken since 6d0a28d98ec0ccd9de0125bf923da7b9a382723f

It is correct that that the program cli option is no longer necessary. But the commit destroys correct buffer delay compensation.

The only lines that should have been removed from `MrsWatson.c`
https://github.com/teragonaudio/MrsWatson/blob/b7886b5a69e32106e6d3fbb88b5fd60a4557e51e/source/MrsWatson.c#L498-L501

All the other logic is still necessary to "automatically" compute tail time (buffer delay compensation) which is then removed at the end.

@nikreiman : Do you want me to make a branch to show what I mean?

**Comment by jarl-dk** [comment]
@nikreiman: You should introduce a regression test for buffer delay compensation based on the code at b599756 (where I believe it works). From there on you can do a `git bisect`. I am pretty confident that it will point at 6d0a28d as the first bad commit.

**Comment by nikreiman** [comment]
@jarl-dk yep, my thoughts exactly. I had a feeling that I had broken delay compensation in 0.9.8, which is possible since as you noted there are no integration or unit tests for this feature. I will introduce tests and fix the bug; this is my mess so I should clean it up. :)

**Comment by jarl-dk** [comment]
@nikreiman : Let me know if you need help or code review...

**Comment by nikreiman** [comment]
@jarl-dk thanks for the offer, I will probably ping you for review when I have a patch ready.

**Comment by fefanto** [comment]
Hi there @jarl-dk @nikreiman. Don't know if this is still relevant, but I recently forked and modified MrsWatson in order to have DelayCompensation "neutralized", as well as having the output file exactly the same length of the input file.
code is here : https://github.com/fefanto/MrsWatson
all I did was adding a "validSamples" field to the sample buffer object, that is updated accordingly to plugin latency (at first read) and to required samples for aligning output to input (at last read).

**Comment by ech2** [comment]
Encountered the same issue, with the latest master
![image](https://user-images.githubusercontent.com/1007225/59506929-1226e900-8edc-11e9-822b-100fe37719ae.png)

**Comment by bafonso** [comment]
I don't believe @fefanto changes ever got merged back to base code. I am unfortunately using OSX (or linux) so I will try to bring in those changes but I also notice that @fefanto switched to VST3 SDK that I do not know if it will have an impact.
